src/base.rs

```diff
@@ -1,10 +1,10 @@
-use std::cmp::min;
-
 use crate::rng::rand;
 use crate::systematic_constants::SYSTEMATIC_INDICES_AND_PARAMETERS;
+use serde::{Deserialize, Serialize};
+use std::cmp::min;
 
 // As defined in section 3.2
-#[derive(Clone, Debug, PartialEq)]
+#[derive(Clone, Debug, Serialize, Deserialize, PartialEq, PartialOrd, Eq, Ord, Hash)]
```

Comment:
I haven't used serde derive before, so I'm not sure exactly how it works. However, I don't think it will do the right thing here. All the types in this file should be serialized using the format specified in RFC6330, which is what the existing methods do.

Reply:
The reason isn't for serializing the data out, its for serializing the data structure itself.  I'm writing a network simulator which is going to use `raptorq` to break messages up into packets.  As a part of the simulator's framework, I make it possible to checkpoint the current state of the simulation so that you can restart the simulation from an earlier checkpoint (the compute cluster I have access to grants you time in fixed-length blocks; if you exceed the time, it aborts the process without sending any signals, so you have to have periodic checkpoints in place to restart from after your process has been aborted).  

That said, you're absolutely right that it won't output bytes in RFC 6330's format; it provides a mechanism to serialize data structures in any number of different formats.  If you haven't used it before, I suggest looking into it.

Reply:
Cool, that sounds like a neat application!

I looked into the `serde` crate, instead of deriving Serialize/Deserialize, can you implement them so that they delegate to the existing serialization methods I have? I think the `serde` methods should still use the RFC serialization format, since it seems like that's a suitable format, and it will avoid surprises where the two serialization formats are different.

Reply:
> I looked into the `serde` crate, instead of deriving Serialize/Deserialize, can you implement them so that they delegate to the existing serialization methods I have?

That would break the serde serialization model, and would be **EXTREMELY** confusing for experienced users.  Serde is enough of a standard in the rust community that it is explicitly mentioned in the [Rust API guidelines](https://rust-lang.github.io/api-guidelines/interoperability.html#c-serde).  I really don't suggest trying to break those rules, you'll get an endless series of bug reports and complaints if you do.

While I agree with you that the existing names make sense (`serialize` and `deserialize`) one breaking change I would suggest is `serialize` -> `to_bytes` per the [guidelines](https://rust-lang.github.io/api-guidelines/naming.html#c-conv), and to remove `deserialize` entirely in favor of implementing [TryFrom](https://doc.rust-lang.org/std/convert/trait.TryFrom.html) on the various types that need it.  This will give you a chance to build in error reporting via the `Result` that is returned.  In addition, the error type should be a [failure](https://crates.io/crates/failure), both because the macros in that crate make it easier to create intelligent errors, and because there is a mechanism in place for [error chaining](https://boats.gitlab.io/failure/fail.html#cause) and [backtraces](https://boats.gitlab.io/failure/fail.html#backtraces) (I suggest reading the [guide](https://boats.gitlab.io/failure/intro.html) to the [failure](https://crates.io/crates/failure) crate, when used well it is a remarkably powerful tool).

Reply:
Yes, creating a custom `Serializer` is the right way to go, **but** you have to be willing to support the rest of `serde`'s data model to make it work right.  That might not be appropriate for `raptorq`.  My suggestion is that instead of using `serde` that way, just create a `to_bytes()` method in the encoder, and implement `TryFrom<&[u8]>` on the decoder.  That will make more sense.